pd/src/consensus/service.rs

```diff
@@ -35,60 +25,32 @@ impl Consensus {
         tokio::spawn(Worker::new(state, queue_rx).await?.run());
 
         Ok(Self {
-            queue: queue_tx,
-            state: State::NoPermit,
-            future: ReusableBoxFuture::new(async { unreachable!() }),
+            queue: PollSender::new(queue_tx),
         })
     }
 }
 
-impl Clone for Consensus {
-    fn clone(&self) -> Self {
-        Self {
-            queue: self.queue.clone(),
-            state: State::NoPermit,
-            future: ReusableBoxFuture::new(async { unreachable!() }),
-        }
-    }
-}
-
 impl tower::Service<ConsensusRequest> for Consensus {
     type Response = ConsensusResponse;
     type Error = BoxError;
     type Future =
         Pin<Box<dyn Future<Output = Result<ConsensusResponse, BoxError>> + Send + 'static>>;
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        match self.state {
-            State::Permit(_) => Poll::Ready(Ok(())),
-            State::NoPermit => {
-                self.future.set(self.queue.clone().reserve_owned());
-                self.state = State::Waiting;
-                self.poll_ready(cx)
-            }
-            State::Waiting => {
-                let permit = ready!(self.future.poll(cx))?;
-                self.state = State::Permit(permit);
-                Poll::Ready(Ok(()))
-            }
-        }
+        self.queue.poll_reserve(cx).map_err(Into::into)
     }
 
     fn call(&mut self, req: ConsensusRequest) -> Self::Future {
-        let permit = if let State::Permit(p) = std::mem::replace(&mut self.state, State::NoPermit) {
-            p
-        } else {
-            panic!("called without poll_ready");
-        };
-
         let span = req.create_span();
         let (tx, rx) = oneshot::channel();
 
-        permit.send(Message {
-            req,
-            rsp_sender: tx,
-            span,
-        });
+        self.queue
```

Comment:
we may want to check if the channel is closed here and bail early (in case it closed between a successful `poll_ready` call and the call to `call`). i don't know how that matters in `pd`, specifically, though, since there's only ever going to be a single consensus background worker and it's expected to live for the entire lifetime of the node...if it's gone, panicking in the `send_item` call is probably also fine.

I didn't change this because the previous code didn't have such a check.